Automattic/jetpack
# Contact Form Block: Displays expected success message when used in FSE template, FSE template part, or widget block

Refactors Contact Form Block generation and processing so that the expected success message displays when the block is used in a FSE template, FSE template part, or block widget.

See https://github.com/Automattic/wp-calypso/issues/60774
See https://github.com/Automattic/wp-calypso/issues/62431

### Proposed changes

The refactor follows a pattern similar to how the Contact Form Widget special-case is handled.

For block templates:
1. `$_wp_current_template_content` is filtered on `template_include` to add a `block_template=canvas` attribute to all contact form blocks in the block template.
2. The `Grunion_Contact_Form` class interprets the `block_template` attribute accordingly when present.
3. The `::process_form_submission()` method properly interprets `contact-form-id=canvas`. It has to hackily reproduce part of `wp-includes/template-loader.php` in order to hydrate `$_wp_current_template_content`(because `$_wp_current_template_content` needs to be rendered in order to populate `Grunion_Contact_Form::$last`.

For block template parts:
1. The `$GLOBALS['grunion_block_template_part_id']` global variable is set when the `wp:core/template-part` block is rendered.
4. If the `$GLOBALS['grunion_block_template_part_id']` is set when the contact form block is rendered, a `block_template_part=<part-id>` attribute is added to the block attributes.
5. The `Grunion_Contact_Form` class interprets the `block_template_part` attribute accordingly when present.
6. The `::process_form_submission()` method properly interprets `contact-form-id=block-template-part-<part-id>`.

For block widgets:
1. `widget_block_content` is filtered to apply `widget=<widget-id>` as an attribute to each block within the block widget.
2. The rest of the existing logic for widgets is reused.

### Does this pull request change what data or activity we track or use?

No.

### Testing instructions

To test the block _template_ integration:

1. Activate a FSE theme (Twenty Twenty-Two is a good one).
2. Add a Contact Form Block to the Single Template.
3. Change "On Submission" to "Show a custom text message". Write a custom text message.
4. Navigate to an individual blog post.
5. Submit the Contact Form Block.
6. Observe your custom success message in the Contact Form Block.
7. Return to Full Site Editor and change your Contact Form Block to use "Show a summary of submitted fields".
8. Navigate to an individual blog post.
9. Submit the Contact Form Block.
10. Observe your submitted fields in the Contact Form block.

<img width="1912" alt="image" src="https://user-images.githubusercontent.com/36432/173824066-1f58da24-dac2-449f-b0ab-41b6da08fe0a.png">

<img width="806" alt="image" src="https://user-images.githubusercontent.com/36432/173823966-7c7e3b9c-c692-409a-acc4-a2db9849740e.png">

To test the block _template part_ integration:

1. Activate a FSE theme (Twenty Twenty-Two is a good one).
2. Add a Contact Form Block to the Footer template part.
3. Change "On Submission" to "Show a custom text message". Write a custom text message.
4. Navigate to the footer on the frontend.
5. Submit the Contact Form Block.
6. Observe your custom success message in the Contact Form Block.
7. Return to Full Site Editor and change your Contact Form Block to use "Show a summary of submitted fields".
8. Navigate to the footer on the frontend.
9. Submit the Contact Form Block.
10. Observe your submitted fields in the Contact Form block.

<img width="1912" alt="image" src="https://user-images.githubusercontent.com/36432/173824199-560380a4-6ee5-470a-8744-b58c15f691ad.png">

<img width="890" alt="image" src="https://user-images.githubusercontent.com/36432/173825798-12f9882f-1204-4076-9189-5a79ea04e850.png">

To test the _widget block_ integration:

1. Activate a legacy theme (Twenty Twenty is a good one).
2. Add a Contact Form Block to a block editor-enabled sidebar.
3. Change "On Submission" to "Show a custom text message". Write a custom text message.
4. Navigate to the sidebar on the frontend.
5. Submit the Contact Form Block.
6. Observe your custom success message in the Contact Form Block.
7. Return to widgets in the backend and change your Contact Form Block to use "Show a summary of submitted fields".
8. Navigate to the sidebar on the frontend.
9. Submit the Contact Form Block.
10. Observe your submitted fields in the Contact Form block.

<img width="1643" alt="image" src="https://user-images.githubusercontent.com/36432/173826924-64ec9e5a-b359-4e68-94e5-c96249003e52.png">

<img width="910" alt="image" src="https://user-images.githubusercontent.com/36432/173827221-fdfbb2d4-5d9a-48ed-b010-bf042be05105.png">
